src/main/java/org/apache/sysml/utils/NativeHelper.java

```diff
@@ -0,0 +1,256 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sysml.utils;
+
+import java.io.IOException;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.util.HashMap;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.File;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.SystemUtils;
+import org.apache.sysml.api.DMLScript;
+import org.apache.sysml.hops.OptimizerUtils;
+
+/**
+ * This class helps in loading native library.
+ * By default, it first tries to load Intel MKL, else tries to load OpenBLAS. 
+ * This behavior can be controlled by setting the environment variable SYSTEMML_BLAS to either mkl or openblas.
+ */
+public class NativeHelper {
+	private static boolean isSystemMLLoaded = false;
+	private static final Log LOG = LogFactory.getLog(NativeHelper.class.getName());
+	private static HashMap<String, String> archMap = new HashMap<String, String>();
+	public static String blasType;
+	static {
+        archMap.put("x86", "x86_32");
+        archMap.put("i386", "x86_32");
+        archMap.put("i486", "x86_32");
+        archMap.put("i586", "x86_32");
+        archMap.put("i686", "x86_32");
+        archMap.put("x86_64", "x86_64");
+        archMap.put("amd64", "x86_64");
+        archMap.put("powerpc", "ppc_64");
+	
+		blasType = getBLASType();
+		if(blasType != null) {
+			try {
+				loadLibrary("systemml", "_" + blasType);
+				isSystemMLLoaded = true;
+				LOG.info("Using native blas: " + blasType);
+			} catch (IOException e) {
+				LOG.warn("Using Java-based BLAS as unable to load native BLAS");
+			}
+		}
+	}
+	
+	private static int maxNumThreads = -1;
+	private static boolean setMaxNumThreads = false;
```

Comment:
please try to remove such global static state that is potentially modified by concurrent threads.